services/github/github-tag.service.js

```diff
@@ -84,39 +76,81 @@ module.exports = class GithubTag extends GithubAuthService {
     return { label: 'tag' }
   }
 
-  static render({ usingSemver, version }) {
+  static render({ version, sort }) {
     return {
       message: addv(version),
-      color: usingSemver ? versionColor(version) : 'blue',
-    }
-  }
-
-  static transform({ usingSemver, includePre, json }) {
-    const versions = json.map(({ name }) => name)
-    if (usingSemver) {
-      return latest(versions, { pre: includePre })
-    } else {
-      return versions[0]
+      color: sort === 'semver' ? versionColor(version) : 'blue',
     }
   }
 
-  async handle({ variant, user, repo }) {
-    const usingSemver = variant !== 'tag-date'
-    const includePre = variant === 'tag-pre'
-
-    const json = await this._requestJson({
+  async fetch({ user, repo }) {
+    return this._requestJson({
       url: `/repos/${user}/${repo}/tags`,
       schema,
-      errorMessages: errorMessagesFor(),
+      errorMessages: errorMessagesFor('repo not found'),
     })
-    if (json.length === 0) {
-      throw new NotFound({ prettyMessage: 'none' })
+  }
+
+  static getLatestTag({ tags, sort, includePrereleases }) {
+    if (sort === 'semver') {
+      return latest(tags.map(tag => tag.name), { pre: includePrereleases })
     }
-    const version = this.constructor.transform({
-      usingSemver,
-      includePre,
-      json,
+    return tags[0].name
+  }
+
+  async handle({ user, repo }, queryParams) {
+    const sort = queryParams.sort
+    const includePrereleases = queryParams.include_prereleases !== undefined
+
+    const tags = await this.fetch({ user, repo })
+    if (tags.length === 0) throw new NotFound({ prettyMessage: 'none' })
+    return this.constructor.render({
+      version: this.constructor.getLatestTag({
+        tags,
+        sort,
+        includePrereleases,
+      }),
+      sort,
     })
-    return this.constructor.render({ usingSemver, version })
   }
 }
+
+const redirects = {
+  GithubTagRedirect: redirector({
+    category: 'version',
+    route: {
+      base: 'github/tag',
+      pattern: ':user/:repo',
+    },
+    transformPath: ({ user, repo }) => `/github/v/tag/${user}/${repo}`,
+    transformQueryParams: params => ({ sort: 'semver' }),
+    dateAdded: new Date('2019-06-25'),
+  }),
+  GithubTagPreRedirect: redirector({
+    category: 'version',
+    route: {
+      base: 'github/tag-pre',
+      pattern: ':user/:repo',
+    },
+    transformPath: ({ user, repo }) => `/github/v/tag/${user}/${repo}`,
+    transformQueryParams: params => ({
+      sort: 'semver',
+      include_prereleases: null,
+    }),
+    dateAdded: new Date('2019-06-25'),
+  }),
+  GithubTagDateRedirect: redirector({
+    category: 'version',
+    route: {
+      base: 'github/tag-date',
+      pattern: ':user/:repo',
+    },
+    transformPath: ({ user, repo }) => `/github/v/tag/${user}/${repo}`,
```

Comment:
Just want to confirm that we don't need the `include_prereleases` query param here correct? If I'm following the original code correctly, `tag-date` excluded pre-releases so this redirect maintains consistency.

Reply:
On the `/releases` endpoint, the `include_prereleases` is meaningful regardless of whether we're sorting by date or semver because there is an "prerelease" boolean.

On the `/tag` endpoint, that doesn't exist - we've just got a string - so `include_prereleases` is only meaningful if `sort=semver` (because we're inferring whether a release is a pre-release ot not based on parsing the string as a semantic version). If `sort=date`, the output is exactly the same regardless of whether you pass `include_prereleases` or not. That's also why there isn't an example for that case.

Instead of `/v/tag/:user/:repo?sort=date` and `/v/tag/:user/:repo?sort=date&include_prereleases` we could set it so that `/v/tag/:user/:repo?sort=date&include_prereleases` throws an error (e.g: "invalid parameter") to make this clearer?

Reply:
Thanks that helps!

I think I'd be inclined to keep the implementation as-is (not throwing an error) given that it's just unused/the default behavior anyway. I could envision someone down the road asking for the most recent, non-prerelease tag by date (not entirely sure what the use case would be, but could definitely see it being requested anyway 😄)

In order to support that, we'd need users to be able to use the `include_prereleases` param to indicate their preference, and could then use it with a slightly updated definition for `getLatestTag`:

```js
static getLatestTag({ tags, sort, includePrereleases }) {
  if (sort === 'semver') {
    return latest(tags.map(tag => tag.name), { pre: includePrereleases })
  }

  if (includePrereleases) {
    return tags[0].name
  }

  // use semver package to find first non-prerelease tag
  const tag = tags.find(t => !semver.prerelease(t.name))
  if (!tag) {
    throw new NotFound({ prettyMessage: 'No release tags found' })
  }

  return tag.name
}
```

Reply:
Also, does the GitHub API for tags automatically sort tags by semver? 

When I take a look at the most recent tags for pytest: https://github.com/pytest-dev/pytest/tags, it seems pretty clear that v4.6.4 was published ~3 hours _after_ the v5.0.0 tag. However, the API response for pytest tags appears to have them sorted by semver:
```json
[
  {
    "name": "5.0.0",
    "zipball_url": "https://api.github.com/repos/pytest-dev/pytest/zipball/5.0.0",
    "tarball_url": "https://api.github.com/repos/pytest-dev/pytest/tarball/5.0.0",
    "commit": {
      "sha": "58bfc7736fc4f88eca669157822e00715c67a9bf",
      "url": "https://api.github.com/repos/pytest-dev/pytest/commits/58bfc7736fc4f88eca669157822e00715c67a9bf"
    },
    "node_id": "MDM6UmVmMzc0ODk1MjU6NS4wLjA="
  },
  {
    "name": "4.6.4",
    "zipball_url": "https://api.github.com/repos/pytest-dev/pytest/zipball/4.6.4",
    "tarball_url": "https://api.github.com/repos/pytest-dev/pytest/tarball/4.6.4",
    "commit": {
      "sha": "d3549df5b94137453b832345a9b8074f518731b0",
      "url": "https://api.github.com/repos/pytest-dev/pytest/commits/d3549df5b94137453b832345a9b8074f518731b0"
    },
    "node_id": "MDM6UmVmMzc0ODk1MjU6NC42LjQ="
  },
  {
    "name": "4.6.3",
    "zipball_url": "https://api.github.com/repos/pytest-dev/pytest/zipball/4.6.3",
    "tarball_url": "https://api.github.com/repos/pytest-dev/pytest/tarball/4.6.3",
    "commit": {
      "sha": "b8e65d03bf563d5de3a285af187fd00a2f202e5e",
      "url": "https://api.github.com/repos/pytest-dev/pytest/commits/b8e65d03bf563d5de3a285af187fd00a2f202e5e"
    },
    "node_id": "MDM6UmVmMzc0ODk1MjU6NC42LjM="
  },
  ...
]
```

And accordingly our badges show v5.0.0 for the latest tag by date:
prod
![](https://img.shields.io/github/tag-date/pytest-dev/pytest.svg)

pr review app
![](https://shields-staging-pr-3610.herokuapp.com/github/v/tag/pytest-dev/pytest.svg)

Reply:
Well, that's an exciting development, isn't it..

Turns out the GitHub API sorts tags... in (reverse) alphabetical order :open_mouth: 

That's not in the API docs anywhere, but after reading that [on stackoverflow](https://stackoverflow.com/a/19615235) (and thinking "wtf?") I set up a test repo just to prove it to myself: https://api.github.com/repos/chris48s/tags-test/tags

```
a24edac  2019-06-29 22:06:10 +0100  (tag: e)
4ca5cd6  2019-06-29 21:59:53 +0100  (tag: f)
bfa97ef  2019-06-29 21:54:52 +0100  (tag: d)
b3614ac  2019-06-29 21:51:20 +0100  (tag: 1)
b645016  2019-06-29 21:51:12 +0100  (tag: 2)
3309766  2019-06-29 21:49:47 +0100  (tag: c)
8d7c78b  2019-06-29 21:49:42 +0100  (tag: b)
5c5a8d3  2019-06-29 21:49:05 +0100  (tag: a)
```

Additionally, the tags page sorts them differently to the API, but also seemingly not chronologically either: https://github.com/chris48s/tags-test/tags

Of course, thinking it through a bit more, @RedSparr0w is actually at least 4 months ahead of me on this: https://github.com/badges/shields/issues/3080#issuecomment-468486956 if not more :laughing: 

Not really sure what to do with this now.. let me have a think about it :D

Reply:
Having done a bit of reading, the only way to order tags by date in a performant way is to use the v4/graphql API. At the moment, we don't have any badges that use the v4 API, but API calls for new GitHub features (e.g: package registry, dependency graph, etc) are only being exposed via the v4 API so we'll need to get into it at some point.

Given that, we do need to look into the v4 API, maybe I'll put this on hold for the moment and use this as an excuse to have a poke at the v4 API and see what the implications would be for the token pool, etc. That might be the solution here, bit its also probably something we need to look into anyway.

Reply:
Sounds like a good plan to me